File: benchmark/scripts/test_easyscf_h5ad_to_rds.py
#!/usr/bin/env python3
"""
easySCF H5AD→RDS 测试 — 步骤1: H5AD → .h5 (自定义格式)

使用 easySCFpy.saveH5() 将标准 H5AD 文件转换为 easySCF 的自定义 .h5 格式。
"""

import time
import traceback
import json
from pathlib import Path

DATA_DIR = Path('/benchmark/data/generated')
TMP_DIR = Path('/benchmark/results/easyscf_h2r_tmp')
TMP_DIR.mkdir(parents=True, exist_ok=True)

# 13 个 H5AD 测试数据集
H5AD_FILES = [
    'scanpy_pbmc3k.h5ad',
    'scanpy_pbmc3k_processed.h5ad',
    'scvelo_dentategyrus.h5ad',
    'scvelo_pancreas.h5ad',
    'cellxgene_pbmc_15k.h5ad',
    'cellxgene_heart_23k.h5ad',
    'cellxgene_brain_40k.h5ad',
    'squidpy_visium_hne.h5ad',
    'squidpy_mibitof.h5ad',
    'squidpy_slideseqv2.h5ad',
    'squidpy_seqfish.h5ad',
    'squidpy_merfish.h5ad',
    'squidpy_imc.h5ad',
]

results = []

for fname in H5AD_FILES:
    h5ad_path = DATA_DIR / fname
    h5_path = TMP_DIR / fname.replace('.h5ad', '_easyscf.h5')
    test_id = fname.replace('.h5ad', '')

    print(f'\n{"="*60}')
    print(f'Testing: {test_id}')
    print(f'  Input:  {h5ad_path}')
    print(f'  Output: {h5_path}')

    if not h5ad_path.exists():
        print(f'  ❌ File not found')
        results.append({'test_id': test_id, 'step1_ok': False, 'step1_error': 'file not found'})
        continue

    try:
        import scanpy as sc
        from easySCFpy import saveH5

        t0 = time.time()
        adata = sc.read_h5ad(str(h5ad_path))
        print(f'  read_h5ad: {adata.n_obs} cells × {adata.n_vars} genes ({time.time()-t0:.1f}s)')

        t1 = time.time()
        saveH5(adata, str(h5_path))
        elapsed = time.time() - t1
        
        ok = h5_path.exists() and h5_path.stat().st_size > 0
        print(f'  saveH5: {"✅" if ok else "❌"} ({elapsed:.1f}s, {h5_path.stat().st_size/1024/1024:.1f} MB)')
        results.append({
            'test_id': test_id,
            'step1_ok': ok,
            'step1_time': round(elapsed, 2),
            'n_cells': adata.n_obs,
            'n_genes': adata.n_vars,
            'h5_size_mb': round(h5_path.stat().st_size / 1024 / 1024, 1),
        })
    except Exception as e:
        print(f'  ❌ Error: {e}')
        traceback.print_exc()
        results.append({
            'test_id': test_id,
            'step1_ok': False,
            'step1_error': str(e),
        })

# 汇总
print(f'\n{"="*60}')
print(f'Step 1 Summary (H5AD → .h5):')
ok_count = sum(1 for r in results if r.get('step1_ok'))
print(f'  {ok_count}/{len(results)} succeeded')
for r in results:
    status = '✅' if r.get('step1_ok') else '❌'
    err = f" — {r.get('step1_error', '')}" if not r.get('step1_ok') else ''
    print(f'  {status} {r["test_id"]}{err}')

# 保存中间结果
out_file = TMP_DIR / 'step1_results.json'
with open(out_file, 'w') as f:
    json.dump(results, f, indent=2)
print(f'\nResults saved to {out_file}')

File: benchmark/scripts/test_easyscf_h5ad_to_rds.R
#!/usr/bin/env Rscript
# easySCF H5AD→RDS 测试 — 步骤2: .h5 (自定义格式) → Seurat → RDS
#
# 前置: 先运行 test_easyscf_h5ad_to_rds.py 生成 .h5 文件

suppressPackageStartupMessages({
  library(easySCFr)
  library(Seurat)
  library(jsonlite)
})

TMP_DIR <- "/benchmark/results/easyscf_h2r_tmp"
RESULTS_FILE <- "/benchmark/results/easyscf_h5ad_to_rds_test.json"

# 读取步骤1结果
step1_file <- file.path(TMP_DIR, "step1_results.json")
if (!file.exists(step1_file)) {
  stop("Step 1 results not found. Run test_easyscf_h5ad_to_rds.py first.")
}
step1 <- fromJSON(step1_file)

results <- list()

for (i in seq_len(nrow(step1))) {
  test_id <- step1$test_id[i]
  step1_ok <- step1$step1_ok[i]
  
  cat(sprintf("\n%s\n", paste(rep("=", 60), collapse = "")))
  cat(sprintf("Testing: %s\n", test_id))
  
  if (!step1_ok) {
    cat("  ⏭️  Step 1 failed, skipping\n")
    results[[length(results) + 1]] <- list(
      test_id = test_id,
      step1_ok = FALSE,
      step2_ok = FALSE,
      step2_error = "step1 failed"
    )
    next
  }
  
  h5_path <- file.path(TMP_DIR, paste0(test_id, "_easyscf.h5"))
  rds_path <- file.path(TMP_DIR, paste0(test_id, "_easyscf.rds"))
  
  if (!file.exists(h5_path)) {
    cat(sprintf("  ❌ .h5 file not found: %s\n", h5_path))
    results[[length(results) + 1]] <- list(
      test_id = test_id,
      step1_ok = TRUE,
      step2_ok = FALSE,
      step2_error = ".h5 file not found"
    )
    next
  }
  
  # 尝试 loadH5 → saveRDS
  tryCatch({
    t0 <- proc.time()["elapsed"]
    obj <- loadH5(h5_path)
    load_time <- proc.time()["elapsed"] - t0
    
    cat(sprintf("  loadH5: ✅ (%s, %.1fs)\n", class(obj)[1], load_time))
    cat(sprintf("  Cells: %d, Genes: %d\n", ncol(obj), nrow(obj)))
    cat(sprintf("  Assays: %s\n", paste(names(obj@assays), collapse = ", ")))
    
    if (length(obj@reductions) > 0) {
      cat(sprintf("  Reductions: %s\n", paste(names(obj@reductions), collapse = ", ")))
    }
    
    t1 <- proc.time()["elapsed"]
    saveRDS(obj, rds_path)
    save_time <- proc.time()["elapsed"] - t1
    
    rds_size <- file.info(rds_path)$size / 1024 / 1024
    cat(sprintf("  saveRDS: ✅ (%.1fs, %.1f MB)\n", save_time, rds_size))
    
    results[[length(results) + 1]] <- list(
      test_id = test_id,
      step1_ok = TRUE,
      step2_ok = TRUE,
      load_time = round(load_time, 2),
      save_time = round(save_time, 2),
      n_cells = ncol(obj),
      n_genes = nrow(obj),
      n_reductions = length(obj@reductions),
      assays = paste(names(obj@assays), collapse = ","),
      rds_size_mb = round(rds_size, 1)
    )
  }, error = function(e) {
    cat(sprintf("  ❌ Error: %s\n", e$message))
    results[[length(results) + 1]] <<- list(
      test_id = test_id,
      step1_ok = TRUE,
      step2_ok = FALSE,
      step2_error = e$message
    )
  })
}

# 汇总
cat(sprintf("\n%s\n", paste(rep("=", 60), collapse = "")))
cat("Step 2 Summary (.h5 → Seurat → RDS):\n")
ok_count <- sum(sapply(results, function(r) isTRUE(r$step2_ok)))
total <- length(results)
cat(sprintf("  %d/%d succeeded\n", ok_count, total))
for (r in results) {
  status <- if (isTRUE(r$step2_ok)) "✅" else "❌"
  err <- if (!isTRUE(r$step2_ok)) paste0(" — ", r$step2_error) else ""
  cat(sprintf("  %s %s%s\n", status, r$test_id, err))
}

# 保存最终结果
write(toJSON(results, auto_unbox = TRUE, pretty = TRUE), RESULTS_FILE)
cat(sprintf("\nResults saved to %s\n", RESULTS_FILE))
